[InstCombine] Propagate poison pow[i], [us]add, [us]sub and [us]mul

badumbatish opened the pull request:
Fixes #146560 as well as propagate poison for [us]add, [us]sub and [us]mul

Commit:
Precommit missed opt for pow poison

Commit:
[InstCombine] Fold poison pow to poison

lukel97 commented:
Is it possible to handle this in ConstantFolding? That way IRBuilder etc. will automatically fold it to poison without the need to run InstSimplify, see: https://github.com/llvm/llvm-project/pull/141821#pullrequestreview-2876118659

lukel97 reviewed (approved):
Ah I see what you mean about having a possibly non-constant arg. In that case I think having it in InstSimplify is fine. 

LGTM but please wait for another reviewer

nikic reviewed (changes requested):
I don't want to add more of these one-off special cases which result in different poison-propagation support between ValueTracking, ConstantFolding and InstSimplify. I've put up https://github.com/llvm/llvm-project/pull/146878 to consolidate ValueTracking and ConstantFolding. We can then do the same for InstSimplify. And then this should just become a matter of adding extra intrinsics to the list.

badumbatish commented on llvm/test/Transforms/InstSimplify/fold-intrinsics.ll:
```diff
@@ -33,7 +33,7 @@ define void @powi(double %V, ptr%P) {
 ; CHECK-LABEL: @powi_i16(
 ; CHECK-NEXT:    store volatile float 1.000000e+00, ptr [[P:%.*]], align 4
-; CHECK-NEXT:    store volatile float [[V:%.*]], ptr [[P]], align 4
+; CHECK-NEXT:    store volatile float [[D:%.*]], ptr [[P]], align 4
```
this comes from the auto updating test tool

Commit:
Refactor support for pow[i] poison folding

Refactor support for pow[i] poison folding by rebasing on Nikic's
changes and implement the changes in InstructionSimplify.

Fix a few test case and added more Intrinsics to switch case in
ValueTracking.

Commit:
Fix naming nits from PR

badumbatish commented:
rebased and refactored. waiting for your pr to be merged.

Commit:
Remove old one off cse of poison folding

Commit:
Added vector support for pow[i] poison

Commit:
Update unittest for pow[i] poison

lukel97 commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6793,6 +6793,9 @@ static Value *simplifyIntrinsic(CallBase *Call, Value *Callee,
   Function *F = cast<Function>(Callee);
   Intrinsic::ID IID = F->getIntrinsicID();
 
+  if (IID != Intrinsic::not_intrinsic && intrinsicPropagatesPoison(IID) &&
+      any_of(Args, IsaPred<PoisonValue>))
+    return PoisonValue::get(F->getReturnType());
```
Do we need this bit here? I thought this PR was just adding support for pow[i]

lukel97 replied:
Ah right, I see it's caused some other intrinsics to be constant folded so we should update the PR title too

badumbatish replied:
updated title as well as head message, ty!

lukel97 commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -7911,7 +7911,52 @@ bool llvm::intrinsicPropagatesPoison(Intrinsic::ID IID) {
+  // case Intrinsic::fptosi_sat:
+  // case Intrinsic::fptoui_sat:
+  // case Intrinsic::lrint:
+  // case Intrinsic::llrint:
```
Leave out the comments for now and just add pow/powi, but keep the list somewhere in your notes so we can return them later.

badumbatish replied:
gotcha, the list will be in [#146769](https://github.com/llvm/llvm-project/issues/146769)

Commit:
Remove case comments for intrinsics poison

nikic reviewed (approved):
LGTM

lukel97 reviewed (approved):
LGTM, I think there's a few fold to poison cases in `simplifyBinaryIntrinsic` can probably be removed afterwards as an NFC
